### PR TITLE
[network] add so_mark as a socket option

### PR DESCRIPTION
**Pull request by klarose**
*Description*: IP Source transparency involves non-local IP addresses being routed as though they were local. This requires some magic in the stack to ensure that those flows are sent back to envoy from the upstream host, rather than back to the original source IP address. We plan on using SO_MARK to do this. So, add it into the socket option factory. My intention is for it to be used by a follow-up PR to #5337. 

This was cherry-picked from PR #5035 where it was already reviewed. I plan on closing that PR.

*Risk Level*: Low. No code invoked in production yet.
*Testing*: Ran newly added UT and other UT in network.
*Docs Changes*: None until we expose this through config.
*Release Notes*: None

### CODE REVIEW

**Comment by klarose** [comment]
Regarding the UT: It's a bit nicer in the original PR, because I had the work from #5351 in it. I'll make that change later as a separate PR, or in this if #5351 gets in quickly enough.

**Comment by klarose** [comment]
@zuercher I'd appreciate it you looked over this again (since it's just a chunk of what you've already reviewed. :) )

File: test/common/network/socket_option_factory_test.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+public:
+  SocketOptionFactoryTest() = default;
+
+  TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls_{[this]() {
+    // Before injecting OsSysCallsImpl, make sure validateIpv{4,6}Supported is called so the static
+    // bool is initialized without requiring to mock ::socket and ::close. :( :(
+    std::make_unique<Address::Ipv4Instance>("1.2.3.4", 5678);
+    std::make_unique<Address::Ipv6Instance>("::1:2:3:4", 5678);
+    return &os_sys_calls_mock_;
+  }()};
```
Comment (zuercher) [inline test/common/network/socket_option_factory_test.cc]: Maybe add a TODO to convert to the simpler test in the future.

File: source/common/network/socket_option_factory.cc [hunk 2]
```diff
@@ -49,6 +49,15 @@ std::unique_ptr<Socket::Options> SocketOptionFactory::buildIpTransparentOptions(
   return options;
 }
 
+std::unique_ptr<Socket::Options> SocketOptionFactory::buildSocketMarkOptions(uint32_t mark) {
+  std::unique_ptr<Socket::Options> options = absl::make_unique<Socket::Options>();
```
Comment (mattklein123) [inline source/common/network/socket_option_factory.cc]: nit: std::make_unique, same elsewhere if applicable.

**Pull request merged**